### PR TITLE
feat(ai): add Cursor provider via agent CLI

### LINKED ISSUE #1004: feat(ai): add Cursor provider via agent CLI
## Parent

Part of #991 — feat(ai): add diff-based code review command (lintro review)

**Phase 9** | Depends on: #995, #996

---

## Summary

Add `CursorProvider` that invokes the Cursor `agent` CLI (`agent --print --output-format json`) so users can run `lintro review` (and other AI features after Phase 14) on a Cursor subscription without Anthropic/OpenAI API keys.

---

## Deliverables

- `lintro/ai/providers/cursor.py` — `BaseAIProvider` implementation
- Register provider in factory / config validation (`provider: cursor`)
- Durable session support (`begin_durable_session` / workspace root from `ReviewContext.repo_root`)
- Unit tests: `tests/unit/ai/test_cursor_provider.py`

---

## Constraints

- **CLI-only** — Cursor has no supported API transport (cursor-sdk attempted in Phase 11 and reverted)
- Requires `agent` on PATH and `CURSOR_API_KEY` or `agent login`

---

## Implementation status

| Field | Value |
|-------|-------|
| Branch | `feat/991-lintro-review` |
| Commit | `f25ef34` |

### PR DESCRIPTION
**Pull request by TurboCoder13**
## Commit Summary (Conventional Commits)

- Title (required, present tense):

  ```text
  feat(ai): add Cursor provider via agent CLI
  ```

- Type:
  - [x] feat (minor)
  - [ ] fix / perf (patch)
  - [ ] docs
  - [ ] refactor
  - [ ] test
  - [ ] chore / ci / style

- Breaking change:
  - [ ] `!` in title or `BREAKING CHANGE:` footer included

## What's Changing

Adds a `CursorProvider` that shells out to the Cursor `agent` CLI so users can run AI features on a Cursor subscription without Anthropic/OpenAI API keys. Registers `AIProvider.CURSOR` in the enum, factory, registry, and model pricing (zero-cost `auto` model).

## Checklist

- [x] Title follows Conventional Commits
- [x] Tests added/updated
- [ ] Docs updated if user-facing
- [x] Local CI passed (`uv run lintro fmt && uv run lintro chk && uv run lintro tst`)

## Closes

- Closes #1004

## Details

- **`lintro/ai/providers/cursor.py`**: invokes `agent --print --output-format json`, parses the JSON envelope, maps auth/timeout/exit errors to shared exceptions, honors caller `timeout` and `max_tokens` (via prompt constraint).
- **Registry wiring**: `AIProvider.CURSOR`, `ProviderInfo`, factory map, and `"auto"` context-window entry.
- **`base.py`**: allow missing API key when `base_url` is set (CLI auth path).
- **Tests**: `tests/unit/ai/test_cursor_provider.py` (flat pytest functions) and `test_cost.py` zero-pricing coverage.

## Summary by CodeRabbit

* **New Features**
  * Added support for a new Cursor-based AI provider option.
  * Expanded model context handling so the automatic model uses a 200,000-token context window.
* **Bug Fixes**
  * Improved API-key handling when a base URL is provided (avoids requiring an API key in that scenario).
  * Updated cost estimation expectations so providers with zero pricing correctly report zero cost.
* **Tests**
  * Added a comprehensive Cursor provider test suite, including response parsing, prompt handling, timeouts, and error cases.

<details><summary><h3>Greptile Summary</h3></summary>

This PR adds a `CursorProvider` that shells out to the Cursor `agent` CLI so users can leverage their Cursor subscription without Anthropic or OpenAI API keys. It wires `AIProvider.CURSOR` through the enum, factory, registry, and model pricing, and includes solid test coverage.

- **`cursor.py`**: Spawns `agent --print --output-format json`, pipes the prompt via stdin, and maps auth/timeout/exit errors to shared exceptions. The `_extract_json_object` fallback (only triggered when `json.loads` fails) and zero-cost `AIResponse` are both correctly implemented.
- **`base.py`**: Adds a fallback that substitutes `\"not-needed\"` when `base_url` is set but no API key is present — a path unused by `CursorProvider` itself (which overrides `_get_client` to return `None`), but it changes error behavior for SDK-based providers configured with a custom `base_url`.
- **Tests**: Good coverage including timeout, auth failure, non-zero exit, invalid JSON, the `is_error` envelope field, and the plain-text-with-braces regression case from the prior review round.
</details>

<h3>Confidence Score: 5/5</h3>

Safe to merge; the Cursor provider is cleanly isolated and previous review issues are resolved.

The new provider is well-contained: it overrides `_get_client` to avoid touching SDK auth, maps every subprocess failure mode to a typed exception, and the `_extract_json_object` fix from the prior round is correctly implemented with a regression test. The two remaining observations — the generic `agent` binary name and the `base.py` dummy-key fallback affecting SDK providers — are both quality concerns rather than broken behavior on the changed path.

`lintro/ai/providers/base.py` — the new `base_url`-without-key branch changes error behavior for Anthropic and OpenAI providers and is worth a second look before it unexpectedly surfaces in those code paths.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| lintro/ai/providers/cursor.py | New CursorProvider implementation using subprocess to call the Cursor `agent` CLI. Error handling, JSON parsing, and prompt construction are well-structured. The `agent` binary name is generic enough that it could resolve to a non-Cursor binary on some machines, and the `--trust` flag is hardcoded with no opt-out. Previous review issues (timeout floor, token cap, `_extract_json_object` truncation) are addressed. |
| lintro/ai/providers/base.py | Adds a `base_url`-without-key fallback that inserts a dummy `"not-needed"` API key. Logically sound for local-proxy use cases but silently changes error behavior for Anthropic/OpenAI providers; `CursorProvider` does not actually use this path since it overrides `_get_client()` entirely. |
| lintro/ai/registry.py | Adds `cursor: ProviderInfo` field and singleton entry with `"auto"` model priced at zero. Consistent with the existing pattern for Anthropic and OpenAI entries. |
| lintro/ai/providers/__init__.py | Registers `CursorProvider` in both the `provider_classes` lookup dict and the `if/elif` dispatch chain. The dual-representation pattern is pre-existing; the Cursor entries are consistent with Anthropic and OpenAI. |
| tests/unit/ai/test_cursor_provider.py | Good coverage of the happy path, error cases (timeout, auth, non-zero exit, invalid JSON, `is_error` flag), and the `_extract_json_object` edge cases including the regression test for plain-text answers with incidental braces. |
| tests/unit/ai/test_cost.py | Updates parametrized cost test to handle zero-cost models correctly; the branched assertion cleanly covers both free and paid models. |
| lintro/ai/model_pricing.py | Adds `"auto": 200_000` to `MODEL_CONTEXT_WINDOWS`. Correct value for a generic large-context model; no naming collision with existing entries. |
| lintro/ai/provider_enum.py | Adds `CURSOR = auto()` to `AIProvider` StrEnum. Minimal, correct change. |

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Caller
    participant CursorProvider
    participant shutil
    participant agent as agent CLI (subprocess)

    Caller->>CursorProvider: __init__()
    CursorProvider->>shutil: which("agent")
    shutil-->>CursorProvider: /usr/local/bin/agent (or None → AINotAvailableError)

    Caller->>CursorProvider: complete(prompt, system, max_tokens, timeout)
    CursorProvider->>CursorProvider: build combined_prompt (system + prompt + token hint)
    CursorProvider->>agent: "subprocess.run([agent, --print, --output-format, json, --trust, --mode, ask, --model, auto], stdin=combined_prompt, timeout=timeout)"
    alt timeout exceeded
        agent-->>CursorProvider: TimeoutExpired
        CursorProvider-->>Caller: AIProviderError("timed out")
    else "returncode != 0"
        agent-->>CursorProvider: stderr with error
        alt "Authentication required" in stderr
            CursorProvider-->>Caller: AIAuthenticationError
        else
            CursorProvider-->>Caller: AIProviderError(returncode + stderr)
        end
    else "returncode == 0"
        agent-->>CursorProvider: stdout JSON envelope
        CursorProvider->>CursorProvider: _parse_json_output(stdout)
        alt "is_error or subtype=="error""
            CursorProvider-->>Caller: AIProviderError
        else valid result
            CursorProvider-->>Caller: "AIResponse(content, tokens, cost=0.0)"
        end
    end
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Caller
    participant CursorProvider
    participant shutil
    participant agent as agent CLI (subprocess)

    Caller->>CursorProvider: __init__()
    CursorProvider->>shutil: which("agent")
    shutil-->>CursorProvider: /usr/local/bin/agent (or None → AINotAvailableError)

    Caller->>CursorProvider: complete(prompt, system, max_tokens, timeout)
    CursorProvider->>CursorProvider: build combined_prompt (system + prompt + token hint)
    CursorProvider->>agent: "subprocess.run([agent, --print, --output-format, json, --trust, --mode, ask, --model, auto], stdin=combined_prompt, timeout=timeout)"
    alt timeout exceeded
        agent-->>CursorProvider: TimeoutExpired
        CursorProvider-->>Caller: AIProviderError("timed out")
    else "returncode != 0"
        agent-->>CursorProvider: stderr with error
        alt "Authentication required" in stderr
            CursorProvider-->>Caller: AIAuthenticationError
        else
            CursorProvider-->>Caller: AIProviderError(returncode + stderr)
        end
    else "returncode == 0"
        agent-->>CursorProvider: stdout JSON envelope
        CursorProvider->>CursorProvider: _parse_json_output(stdout)
        alt "is_error or subtype=="error""
            CursorProvider-->>Caller: AIProviderError
        else valid result
            CursorProvider-->>Caller: "AIResponse(content, tokens, cost=0.0)"
        end
    end
```

</a>
</details>

<sub>Reviews (10): Last reviewed commit: ["style(test): address CodeRabbit PT019 an..."](https://github.com/lgtm-hq/py-lintro/commit/b6d6ed70c143547e67877bd4053bb0ccb429ed79) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=39939256)</sub>

### CODE REVIEW

File: lintro/ai/providers/cursor.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+                input=combined_prompt,
+                capture_output=True,
+                text=True,
+                timeout=max(timeout, 300.0),
+                check=False,
+            )
+        except subprocess.TimeoutExpired as e:
+            raise AIProviderError(
+                f"Cursor CLI timed out after {timeout}s",
+            ) from e
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/ai/providers/cursor.py]: Fixed in dc1cf899: `subprocess.run` now receives the caller's `timeout` directly (removed `max(timeout, 300.0)`). Added `test_timeout_passed_to_subprocess`.

File: lintro/ai/providers/cursor.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+                f"Full response: {json.dumps(data)[:1000]}",
+            )
+
+        # The agent CLI may prefix the JSON with prose. Extract the
+        # outermost JSON object so downstream parsers don't choke.
+        content = self._extract_json_object(content)
+
+        usage = data.get("usage", {})
+        input_tokens = usage.get("inputTokens", 0)
+        output_tokens = usage.get("outputTokens", 0)
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/ai/providers/cursor.py:261]: Acknowledged — `agent --output-format json` does not expose token counts in the envelope (confirmed via Cursor docs). Zeros are intentional; documented at line 248 with a comment. Tests use synthetic `usage` keys only to verify parsing when present.

File: lintro/ai/providers/cursor.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        content = data.get("result", "")
+        if not content:
+            logger.warning(
+                f"Cursor CLI returned empty result. "
+                f"Full response: {json.dumps(data)[:1000]}",
+            )
+
+        # The agent CLI may prefix the JSON with prose. Extract the
+        # outermost JSON object so downstream parsers don't choke.
+        content = self._extract_json_object(content)
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/ai/providers/cursor.py]: Fixed in 6f950a8e: `_extract_json_object` is now only applied when `json.loads(content)` fails, so plain-text answers with incidental braces pass through verbatim. Added regression test.

File: tests/unit/ai/test_cursor_provider.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+def test_cursor_provider_default_model(provider):
+    """Use auto as the default model."""
+    assert_that(provider.model_name).is_equal_to("auto")
+
+
+def test_cursor_provider_custom_model(_mock_agent_on_path):
+    """Accept a custom model override."""
+    p = CursorProvider(model="claude-opus-4-8-thinking-high")
+    assert_that(p.model_name).is_equal_to("claude-opus-4-8-thinking-high")
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply tests/unit/ai/test_cursor_provider.py]: Fixed in b6d6ed70.

File: tests/unit/ai/test_cursor_provider.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+
+
+def test_complete_raises_on_subprocess_timeout(provider):
+    """Raise AIProviderError when CLI times out."""
+    with patch(
+        "subprocess.run",
+        side_effect=subprocess.TimeoutExpired(cmd="agent", timeout=60),
+    ):
+        with pytest.raises(AIProviderError, match="timed out"):
+            provider.complete("Hello")
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply tests/unit/ai/test_cursor_provider.py]: Fixed in b6d6ed70.
